tlatools/org.lamport.tlatools/src/util/ExecutionStatisticsCollector.java

```diff
@@ -60,75 +64,131 @@ public String toString() {
 		// of its state if the VM decides to shutdown (e.g. because TLC is done).
 		final Thread thread = new Thread(new Runnable() {
 			@Override
 			public void run() {
-				if (isEnabled()) {
-					// Include identifier to track individual installations (not users!).
-					parameters.put("id", getIdentifier());
-					submit(parameters, dontWaitForCompletion);
-				}
+				collect0(parameters);
 			}
+
 		}, "TLC Execution Statistics Collector");
 		thread.setDaemon(dontWaitForCompletion);
 		thread.start();
 	}
 	
-	/*
-	 * file == ~/.tlaplus/esc.txt
-	 * fl == first line of file interpreted as a string without terminal chars
-	 * in/out == opt-in & opt-out
-	 * y/r/n == data collected with constant id/data collected with random id/data not collected
-	 * 
-	 *       | No file | fl unreadable | fl empty | fl = "NO_UDC" | fl = "RANDOM_IDENTIFIER" | fl any other string
-	 * ==========================================================================================================
-	 * | out |   y     |       n       |    n     |       n       |            r             |         y         |
-	 * ----------------------------------------------------------------------------------------------------------
-	 * | in  |   n     |       n       |    n     |       n       |            r             |         y         |
-	 * ----------------------------------------------------------------------------------------------------------
-	 */
+	protected void collect0(final Map<String, String> parameters) {
+		/*
+		 * | `esc.txt`                            | DNS Query | DNS Query Succeeds (Private Reporting) | DNS Query NXDOMAIN (Public Reporting) |
+		 * |--------------------------------------|-----------|----------------------------------------|---------------------------------------|
+		 * | `NO_STATISTICS`                      | n         | N/A                                    | N/A                                   |
+		 * | No `esc.txt`, unreadable, or empty   | y         | UUIDv1                                 | N/A                                   |
+		 * | `RANDOM_IDENTIFIER`                  | y         | UUIDv4                                 | UUIDv4                                |
+		 * | Some string `S`                      | y         | `S`                                    | `S`                                   |
+		 * |--------------------------------------|-----------|----------------------------------------|---------------------------------------|
+		 */
+		String line = null;
+
+		// Check if the installation explicitly requested opt-out by having created the
+		// file containing the NO_ESC_STR.
+		final File udcFile = new File(pathname);
+		if (udcFile.exists()) {
+			try (BufferedReader br = new BufferedReader(new FileReader(udcFile))) {
+				line = br.readLine();
+				if (NO_ESC_STR.equals(line.trim())) {
+					return;
+				}
+			} catch (Exception swallow) {
+			}
```

Comment:
If the file is unreadable for some unknown reason, it might be better to return or throw instead of silently proceeding. (Only a successful read of the file can prove that it _doesn't_ contain `NO_ESC_STR`.)

Maybe something like
```java
try (...) {
   ...
} catch (FileNotFoundException | NoSuchFileException swallow) {
    // file does not exist; continue because the user has not expressed a preference
} catch (IOException e) {
    throw new RuntimeException("Unable to read " + udcFile, e);
}
```

Reply:
The idea is to not bother the user with exec stats even if something is amiss.  For example, we would have to think about it a (TLC)RuntimeException interferes with the REPL, model-checking, or simulation. I’m open to changing the default to disable execution statistics in both cases if the file is unreadable.

Reply:
https://github.com/tlaplus/tlaplus/pull/1172/commits/a410fdf9c35743e28a63bf48ba1874d0dad58fce